### DISCUSSION TITLE
No SMTP servers were available for `mydomain.com`. No hosts to try.

### DISCUSSION THREAD

**Discussion by Salvogiarracca** [postalserver/postal#3252] [Help with using Postal] [answered]
Hello everyone, I have some problems setting up `Postal` as `SMTP relay`. 

A bit of context:
I bought a VPS and I installed Postal on it, configured dns entries as the guide suggests, and finally having a `10/10` score on `mail-tester.com`. Then I setup a mail server on my private home server (IMAP/POP3) that let me store mailboxes privately. The accounts on this mail server are on the form `test@mydomain.com`. Also I installed a `wireguard` VPN on my private server in order to have VPS (10.0.1.2) and MailServer(10.0.1.1) on the same LAN(10.0.1.0/24). All the services runs in docker containers with `network_mode: host`, except for Postal that is in its default configuration. Up to here no problem at all, each container can reach all the others (ping, telnet).

The problem:
- Created a new `SMTP endpoint`, with hostname myserver.com (the mailserver), port `25`, ssl `auto`.
- Created a new route `test@mydomain.com` associated to the previously created SMTP endpoint (myserver.com).

I suppose that this should be translated like this: "When I (postal) receive an email with the `TO` field equal to `test@mydomain.com`, i redirect it to the SMTP endpoint (myserver.com)". Sad story, it does not work.

The only useful information from UI is a strange `SOFT FAIL`: 
`No SMTP servers were available for mydomain.com. No hosts to try.`

I don't have DNS problems because, as said before, each container can reach the others. I use a FQDN for the mailserver (mapped in `/etc/hosts` file). I don't think it should be a problem related to the mailserver because `telnet` command at port 25 works, but it is like the routing rule is not considered at all. I tried to add `smtp_relays` in the `postal.yml`, although I didn't understand how this is related to the `SMTP endpoint` in the UI. Even the option `use_local_ns_for_domain_verification: true` did not make any change.

I hope that I have been clear, any suggestion/help to fix the problem would be very appreciated.

**Comment by willpower232** [comment]
> No hosts to try.

This is the biggest clue, I think this means that either there are no working MX records for the domain you're trying to send to or your server provider is blocking your SMTP connections outbound.

https://github.com/postalserver/postal/blob/fd3c7ccdf6dc4ee0a76c9523cbd735159e4b8000/app/senders/smtp_sender.rb#L46-L47

**Reply by Salvogiarracca** [reply]
Thank you for the response. Actually the domain configuration in Postal looks fine. `SPF`, `DKIM`, `Return Path` and `MX` records are green. The problem is between `Postal` and the `Mail Server`. Indeed, I'm able to do `telnet myserver.com 25` from the VPS (even inside the postal containers), but still `No hosts to try`.

**Reply by willpower232** [reply]
Can your Postal server send any email?

I'm presuming that `myserver.com` is not your Postal server, is that correct?

**Reply by Salvogiarracca** [reply]
Yes, Postal can send emails. `myserver.com` is my private mail server, and both are connected through a VPN. I wrote an entry in `/etc/hosts` (10.0.1.1  myserver.com) on the VPS (where postal is) in order to be resolved correctly.

**Reply by willpower232** [reply]
Ah well what happens if you enter the IP address into Postal instead of myserver.com?

**Reply by Salvogiarracca** [reply]
Unfortunately, nothing changes. At the beginning I used IP instead of `myserver.com` but I saw on other discussions that is better to use FQDN.

**Reply by willpower232** [reply]
did you try telnet from the Postal server or somewhere else?

**Reply by Salvogiarracca** [reply]
`mydomain.com` is reachable from both the `postal-smtp` and `postal-worker` containers. I executed `telnet myserver.com 25` from within both containers, and it works.

**Reply by willpower232** [reply]
Well I'm afraid I'm out of ideas, everything sounds grand and working but there must be something wrong somewhere.

I guess you could edit the smtp_sender.rb file I linked above to include some more logs about what, if any, endpoints are actually tried and maybe that will help explain what isn't happening.

**Reply by Salvogiarracca** [reply]
I tried to replicate the setup in a virtualized environment: the email is correctly forwarded to the mail server. The only difference between the two is that there isn't a VPN in the virtual setup between Postal (`relay.local` at `10.0.3.1`) and the mailserver (`mailserver.local` at `10.0.3.2`). I was wondering, can the VPN tunnel (`wireguard` uses an `UDP`) port be a problem?

**Reply by willpower232** [reply]
Ah potentially, I'm not really familiar with the low level network specifics but at this point, I imagine the problem will be something complicated like that.

**Reply by Salvogiarracca** [reply]
I will check that. Can you tell me how the `SMTP endpoints` (UI) and the `smtp_relays` (postal.yml) are related? I tried to add manually the smtp servers under the `postal` section in this way:
`smtp_relays: [ 'smtp://myserver.com:25?ssl_mode=auto', 'smtp://10.0.1.1:25?ssl_mode=auto' ]`
I assume it is syntactically correct because postal does not complain when restarted.

**Reply by willpower232** [reply]
If you have relays added to the config file, Postal will use them to send all outgoing messages so might actually be useful for your use case.

Endpoints in the UI are for passing on some email based on the rest of the route. You can also have non-SMTP endpoints to direct the message body at a HTTP API for example.

**Reply by Salvogiarracca** [reply, marked as answer]
I solved the issue. The problem was with DNS resolution. While `/etc/hosts` was correctly configured and `telnet` could reach `myserver.com`, it seems that `Postal` does not handle IP addresses well (the `endpoint` variable was empty). I decided to use a public FQDN instead, and the logs showed that `Postal` attempted to reach that host (unsuccessfully, as it was not an SMTP server). However, this time, the `endpoint` variable was printed correctly. 

To address the issue, I installed `dnsmasq` and configured `systemd-resolved` to use `dnsmasq` as the upstream server. This setup ensures that only the DNS queries for the VPN domain are sent to `dnsmasq`.

Thank you for the support!

**Reply by willpower232** [reply]
Ah good stuff, the hosts file does behave in mysterious ways but then I guess its a little too easy to use!
